flink-connector-aws-dynamodb/src/main/java/org/apache/flink/streaming/connectors/dynamodb/sink/DynamoDbSinkElementConverter.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.dynamodb.sink;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.connector.sink2.SinkWriter;
+import org.apache.flink.connector.base.sink.writer.ElementConverter;
+import org.apache.flink.streaming.connectors.dynamodb.util.DynamoDbAttributeValueUtils;
+import org.apache.flink.util.Preconditions;
+
+import software.amazon.awssdk.services.dynamodb.model.WriteRequest;
+
+/**
+ * An implementation of the {@link ElementConverter} that uses the AWS DynamoDb SDK v2. The user
+ * needs to provide a {@link DynamoDbRequestConverter} of the {@code InputT} to transform it into a
+ * {@link DynamoDbWriteRequest} that may be persisted.
+ */
+@Internal
+public class DynamoDbSinkElementConverter<InputT>
```

Comment:
It looks like we have created wrappers around the AWS SDK DynamoDB model within the sink. Can you please explain your reasoning behind this approach? The idea behind hiding the ElementConverter is to 1/ hide the internal implementation and 2/ enable sink to manage de/serialisation. However I am struggling to see the benefits here since the 2x models are very similar. If we will copy the underlying model it might be best to use that model, since users might already be creating these objects. 

Did you consider using a more generic `ElementConverter`, like the [DynamoDBMapper](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBMapper.html)? I am familiar with this so not sure if there are performance/feature limitations

Reply:
We had a discussion earlier on that matter [here](https://github.com/apache/flink/pull/18518#discussion_r793711778)

If we go for hiding the converter we do not leak underlying dynamodb sdk classes to the user, but we have to create a wrapper around their model. This means if the dynamodb client changes over time we will have to propagate model changes too. 

I think it may be better to allow the user to define the element converter and not do any translation to write the request ourselves (like we had it before this PR: https://github.com/YuriGusev/flink/commit/de83346dc24367ff1101ae13934dfcf1fc649de2#diff-1c18ee5b3c5b4d827778ffdf4e5d2219dbcb02fc45be0d3665357633a134995c).

Reply:
If we revert the change, we must ensure that if the dynamo client classes change over time we update our de/serialiser. Otherwise we might get in the situation where user `ElementConverter` generates classes that are not supported by our serialiser. The impact would be that we lose/break attributes when restoring from checkpoints. We can mitigate this by calling it out in the docs/PR template as something to check when updating the SDK.

> I think it may be better to allow the user to define the element converter and not do any translation ourselves (like we had it before this PR

Given we cannot escape these wrappers, I agree with this.

Reply:
Agree that having a wrapper provides limited benefits in this case and that it makes more sense to expose the DynamoDB SDK classes directly. 

To mitigate the risk of not updating the de/serializers when doing a SDK update, it would be great if we could find a way to programmatically verify during compile time that the SDK is still compatible with the de/serializers. But I'm not sure how much work that would required and if it's worth the effort.

Reply:
Thanks @sthm good comment
If I understand correcty, we are concerned that the sink uses version x of dynamodb sdk, while the user is using version y, and the versions are incompatible for some reason? I think during compile time it would be hard to detect, as this is somewhat a dependency management issue - unless we shade the aws sdk and only allow to pass these shaded objects as an input?
Otherwise, do we think an error during sink startup makes sense that the sdk classes used are incompatible with the one used in the sink due to serialization issues? similar to how flink fails if an object that is passed to the operators is not serializable

Reply:
@dannycranmer makes sense, I can go ahead with implementing that if you agree @YuriGusev 
regarding @sthm comment on serialization, do we want to shade the dynamodb sdk? is it something the community was considering?

Reply:
I would rather avoid shading the SDK. It means that end users cannot update the SDK without building from source, this has caused issues before. For example, there is no way to add support for new AWS regions for older versions of the connector without building from source. On top of this, if you include multiple connectors, you end up with multiple versions of the SDK bloating the size of the jar. 

We can think about a build time check as a follow up https://issues.apache.org/jira/browse/FLINK-29688